airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py

```diff
@@ -43,8 +43,12 @@ def __init__(self, email: str, password: str):
 
 
 class SourceZendeskSupport(YamlDeclarativeSource):
-    def __init__(self):
-        super().__init__(**{"path_to_yaml": "manifest.yaml"})
+    def __init__(self, catalog: Optional[ConfiguredAirbyteCatalog], config: Optional[Mapping[str, Any]], state: TState, **kwargs):
+        # Before 2024-11-11, the config was being modified in `streams`. We can't do that anymore because `ConcurrentDeclarativeSource` use
+        # the config to make concurrent components. Hence, the "main" config needs to be the declarative one and if Python sources need
+        # something a bit different, it needs to deduce it from the declarative config
```

Comment:
Without those lines, we would get the following error:
`` 
{
    "type": "TRACE",
    "trace": {
        "type": "ERROR",
        "emitted_at": 1731074703862,
        "error": {
            "message": "Error starting the sync. This could be due to an invalid configuration or catalog. Please contact Support for assistance. Error: time data 'None' does not match format '%Y-%m-%dT%H:%M:%SZ'",
            "stack_trace": "Traceback (most recent call last):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py\", line 22, in _get_source
    return SourceZendeskSupport(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py\", line 47, in __init__
    super().__init__(catalog=catalog, config=config, state=state, **{\"path_to_yaml\": \"manifest.yaml\"})
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/yaml_declarative_source.py\", line 31, in __init__
    super().__init__(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/concurrent_declarative_source.py\", line 65, in __init__
    self._concurrent_streams, self._synchronous_streams = self._group_streams(config=config or {})
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/concurrent_declarative_source.py\", line 164, in _group_streams
    cursor, connector_state_converter = self._constructor.create_concurrent_cursor_from_datetime_based_cursor(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 564, in create_concurrent_cursor_from_datetime_based_cursor
    start_date = interpolated_start_date.get_datetime(config=config)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/datetime/min_max_datetime.py\", line 58, in get_datetime
    time = self._parser.parse(str(self.datetime.eval(config, **additional_parameters)), datetime_format)  # type: ignore # datetime is always cast to an interpolated string
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/datetime/datetime_parser.py\", line 35, in parse
    parsed_datetime = datetime.datetime.strptime(str(date), format)
  File \"/Users/maxime/.pyenv/versions/3.10.11/lib/python3.10/_strptime.py\", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File \"/Users/maxime/.pyenv/versions/3.10.11/lib/python3.10/_strptime.py\", line 349, in _strptime
    raise ValueError(\"time data %r does not match format %r\" %
ValueError: time data 'None' does not match format '%Y-%m-%dT%H:%M:%SZ'
"
        }
    }
}
```

The reason is that the `start_date` is optional in the config but the child stream (`SourceZendeskSupport` in this case) modify the config to instantiate the declarative streams, `ConcurrentDeclarativeSource` do not see those changes and this might cause errors when grouping the streams as part of `ConcurrentDeclarativeSource.__init__`.

Reply:
What still isn't obvious to me is why we have to inject the optional 2 year start date in code.

I'm assuming that this is failing for the semi incremental cursor here:
```
incremental_sync:
  type: DatetimeBasedCursor
  cursor_datetime_formats:
    - "%Y-%m-%dT%H:%M:%SZ"
    - "%Y-%m-%dT%H:%M:%S%z"
  datetime_format: "%Y-%m-%dT%H:%M:%SZ"
  cursor_field: "{{ parameters.get('cursor_field',  'updated_at') }}"
  start_datetime:
    datetime: "{{ config.get('start_date') }}"
```

I feel like we should be able to just add a condition to the interpolation to look back 2 years which simplify the python code that's required. Something like this. I'll experiment with why we didn't just do this instead of this extra config transformation logic

```
start_datetime:
  datetime: "{{ config.get('start_date') or day_delta(-730, '%Y-%m-%dT%H:%M:%SZ') }}"

```

Reply:
noting that i removed this block of code because we can actually account for this using interpolation instead of custom python code:

in `semi_incremental_stream.incremental_stream`:
`"{{ config.get('start_date') or day_delta(-730, '%Y-%m-%dT%H:%M:%SZ') }}"`

If `start_date` is missing, then we'll default to two years previously, and i tested this removing it from a config and the sync will not experience this error